Shared/mods/deathmatch/logic/luadefs/CLuaElementDefsShared.cpp

```diff
@@ -1,87 +1,78 @@
+//  bool setElementData ( element theElement, string key, var value, [bool synchronize = true / string syncType = "broadcast"] )
+bool CLuaElementDefs::setElementData(lua_State* const luaVM, CElement* const element, std::string key, CLuaArgument newValue, const std::optional<std::variant<bool, ESyncType>> optionalNewSyncType)
 #endif
-            if (pVariable)
-            {
-                pVariable->Push(luaVM);
-                return 1;
-            }
-        }
+{
+#ifndef MTA_CLIENT
+    ESyncType newSyncType = ESyncType::BROADCAST;
+    if (optionalNewSyncType.has_value())
+    {
+        const auto& value = optionalNewSyncType.value();
+        if (std::holds_alternative<bool>(value) && !std::get<bool>(value))
+            newSyncType = ESyncType::LOCAL;
+        else
+            newSyncType = std::get<ESyncType>(value);
     }
-    else
-        m_pScriptDebugging->LogCustom(luaVM, argStream.GetFullErrorMessage());
 
-    // Failed
-    lua_pushboolean(luaVM, false);
-    return 1;
-}
+    LogWarningIfPlayerHasNotJoinedYet(luaVM, element);
+#endif
+    EDATA_KEY_WARN_TRUNCATE(key);
 
-int CLuaElementDefs::HasElementData(lua_State* luaVM)
-{
-    //  bool hasElementData ( element theElement, string key [, bool inherit = true ] )
 
 #ifdef MTA_CLIENT
-    CClientEntity* pElement;
+    return CStaticFunctionDefinitions::SetElementData(*element, key.c_str(), newValue, optionalIsSynced.value_or(true));
 #else
-    CElement* pElement;
+    return CStaticFunctionDefinitions::SetElementData(element, key.c_str(), newValue, newSyncType);
 #endif
-    SString   strKey;
-    bool      bInherit;
+}
 
-    CScriptArgReader argStream(luaVM);
-    argStream.ReadUserData(pElement);
-    argStream.ReadString(strKey);
-    argStream.ReadBool(bInherit, true);
 
-    if (argStream.HasErrors())
-    {
-        return luaL_error(luaVM, argStream.GetFullErrorMessage());
-    }
+#ifdef MTA_CLIENT
+bool CLuaElementDefs::HasElementData(lua_State* const luaVM, CClientEntity* const element, std::string key, const std::optional<bool> inherit)
+#else
+bool CLuaElementDefs::hasElementData(lua_State* const luaVM, CElement* const element, std::string key, const std::optional<bool> inherit)
+#endif
+{
+    EDATA_KEY_WARN_TRUNCATE(key);
 
-    if (strKey.length() > MAX_CUSTOMDATA_NAME_LENGTH)
-    {
-        // Warn and truncate if key is too long
-        m_pScriptDebugging->LogCustom(luaVM, SString("Truncated argument @ '%s' [%s]", lua_tostring(luaVM, lua_upvalueindex(1)),
-                                                     *SString("string length reduced to %d characters at argument 2", MAX_CUSTOMDATA_NAME_LENGTH)));
-        strKey = strKey.Left(MAX_CUSTOMDATA_NAME_LENGTH);
-    }
+#ifdef MTA_CLIENT
+    return element->GetCustomData(key.c_str(), inherit.value_or(true)) != nullptr;
+#else
+    return CStaticFunctionDefinitions::GetElementData(element, key.c_str(), inherit.value_or(true)) != nullptr;
```

Comment:
`inherit.value_or(true)`

very cool! didn't know you could do this

Reply:
Yeah, indeed. I went thru the doc of `optional` and `variant`. They're pretty clever things.